executor/importer/chunk_process.go

```diff
@@ -256,10 +218,128 @@ func (p *chunkProcessor) encodeLoop(ctx context.Context) error {
 	return nil
 }
 
-func (p *chunkProcessor) deliverLoop(ctx context.Context) error {
+// ChunkProcessor is used to process a chunk of data, include encode data to KV
+// and deliver KV to local or global storage.
+type ChunkProcessor interface {
+	Process(ctx context.Context) error
+}
+
+type baseChunkProcessor struct {
+	enc         *chunkEncoder
+	logger      *zap.Logger
+	kvCodec     tikv.Codec
+	deliverLoop func(ctx context.Context) error
+	encodeDone  func(ctx context.Context)
+
+	// initialized when Process
+	metrics *metric.Common
+
+	checksum        verify.KVChecksum
+	deliverTotalDur time.Duration
+}
+
+func (p *baseChunkProcessor) Process(ctx context.Context) (err error) {
+	task := log.BeginTask(p.logger, "process chunk")
+	defer func() {
+		task.End(zap.ErrorLevel, err,
+			zap.Duration("readDur", p.enc.readTotalDur),
+			zap.Duration("encodeDur", p.enc.encodeTotalDur),
+			zap.Duration("deliverDur", p.deliverTotalDur),
+			zap.Object("checksum", &p.checksum),
+		)
+		if err == nil && p.metrics != nil {
+			p.metrics.ChunkCounter.WithLabelValues(metric.ChunkStateFinished).Inc()
+		}
+	}()
+	if err2 := p.enc.initProgress(); err2 != nil {
+		return err2
+	}
+
+	if metrics, ok := metric.GetCommonMetric(ctx); ok {
+		p.enc.metrics = metrics
+		p.metrics = p.enc.metrics
+	}
+
+	group, gCtx := errgroup.WithContext(ctx)
+	group.Go(func() error {
+		return p.deliverLoop(gCtx)
+	})
+	group.Go(func() error {
+		defer p.encodeDone(gCtx)
+		return p.enc.encodeLoop(gCtx)
```

Comment:
I slightly prefer we use a local variable `kvsCh`, so it's more obvious how these two loops interact with each other and how closing channel represents one loop is done . However it will make encapsulation (like `sendFn`) more difficult. LGTM